shell/preprocessor.go

```diff
@@ -0,0 +1,60 @@
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/anmitsu/go-shlex"
+	"github.com/mgood/go-posix"
+)
+
+// Command Preprocessor for shell input
+// Handles things like glob expansion ('*' and '?'),
+// environment variables, and shell variables
+
+func preprocess(input string) ([]string, error) {
+	args, err := shlex.Split(input, true)
+
+	mapping := envMap()
+
+	result := []string{}
+	for _, arg := range args {
+
+		if strings.ContainsAny(arg, "*?[") {
+			matches, err := filepath.Glob(arg)
+			if err != nil {
+				return result, err
+			}
+			// add expanded glob into result
+			for _, match := range matches {
+				result = append(result, match)
+			}
+			continue
+		}
+
+		if arg[0] == '$' {
+			val, err := posix.Expand(arg, mapping)
```

Comment:
Oh I meant to look at this sooner, but I believe you can use `ExpandEnv` here instead to handle the env var substitution

Reply:
Oh! I'll make another PR to address that. Thanks @mgood

Reply:
I think in this case we want control over where the environment vars are coming from, so passing in `env` would be preferable.